bin/ch-test

```diff
@@ -0,0 +1,322 @@
+        fi
+        bl=$(command -v docker)
+        bv=$(docker --version | awk '{print $3}' | sed -e 's/,//')
+        ;;
+    *)
+        fatal "builder: $CH_BUILDER: not supported"
+        ;;
+    esac
+    printf '%s %s: OK\n\n' "$bl" "$bv"
+    check_min_ver 'builder' "$min" "$bv"
+}
+
+check_min_ver () {
+    if [[ $(printf '%s\n%s\n' "$2" "$3" | sort -V | head -n1) != "$2" ]]; then
+        fatal "$1: mininum version $2"
+    fi
+}
+
+check_scope () {
+    case $1 in
+    quick|standard|full)
+        return 0
+        ;;
+    *)
+        fatal "scope: invalid: $1"
+        ;;
+    esac
+}
+
+check_tardir () {
+    printf '\nchecking tarball directory: %s: ' "$CH_TEST_TARDIR"
+    # The run phase requires artifacts from a successful build phase. Thus,
+    # we check sanity based on the minimal set of artifacts (quick scope).
+    check_tardir_file alpine39{.tar.gz,.sqfs,.pq_missing}
+    check_tardir_file alpine39_dp{.tar.gz,.sqfs,.pq_missing}
+    check_tardir_file chtest{.tar.gz,.sqfs,.pq_missing}
+    printf 'OK\n'
+}
+
+check_tardir_file () {
+    local missing
+    for f in "$@"; do
+        if [[ -f ${CH_TEST_TARDIR}/${f} ]]; then
+            return 0
+        else
+            missing+=("${CH_TEST_TARDIR}/${f}")
+        fi
+    done
+    fatal "phase: $phase: missing build phase artifact(s): ${missing[*]}"
+}
+
+fatal () {
+    printf 'error: %s\n\n' "$1" 1>&2
+    exit 1
+}
+
+# Create test directories for TAR and IMG. This is to avoid cryptic bats error
+# (#347) that occurs when setting a CH_TEST_IMGDIR or CH_TEST_TARDIR variables
+# to a path where there are two or more subdirectories that do not exist.
+mk_dirs () {
+if [[ ! -d $CH_TEST_TARDIR ]]; then
+    echo "creating $CH_TEST_TARDIR"
+    mkdir -p "$CH_TEST_TARDIR"
+fi
+if [[ ! -d $CH_TEST_IMGDIR ]]; then
+    echo "creating $CH_TEST_IMGDIR"
+    mkdir -p "$CH_TEST_IMGDIR"
+fi
+}
+
+# Remove CH_TEST_PERMDIRS directories. Ensure that the target directory has
+# only two sub directories, e.g., nopass and pass.
+rm_ () {
+    if [[ $(find "${d}" -maxdepth 1 -mindepth 1 | wc -l) == 2 ]] && \
+       (find "${d}" -maxdepth 1 -mindepth 1 -name pass -type d) && \
+       (find "${d}" -maxdepth 1 -mindepth 1 -name nopass -type d); then
+        echo "removing ${d}"
+        sudo rm -rf --one-file-system "$d"
+    fi
+}
+
+single_assignment () {
+    if [[ -n $1 ]]; then
+        fatal "phase: assignment more than once (originally set to: $phase)"
+    fi
+}
+
+set_builder () {
+    if [[ -n $builder ]]; then
+        export CH_BUILDER=$builder
+        method='command line'
+    elif [[ -n $CH_BUILDER ]]; then
+        method='environment variable'
+    else
+        builder_choose
+        method='default'
+    fi
+    printf 'setting builder to %s (%s)\n' "$CH_BUILDER" "$method"
+    if [[ $CH_BUILDER == 'ch-grow' ]]; then
+        set_chvar CH_GROW_STORAGE 'ch-grow storage' '' "$CH_GROW_STORAGE" /var/tmp/ch-grow_
+    fi
+}
```

Comment:
I still feel like there's probably a better way to do this than have a function specific for a single variable, but also that you've probably put enough work into it and we should leave it alone for now.

Reply:
> I still feel like there's probably a better way to do this than have a function specific for a single variable...

I completely agree. Unfortunately I haven't been able to come up with anything better. Would you like to take a shot at it?